src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -8333,6 +8376,280 @@ private StatementSyntax ParseLocalDeclarationStatement()
+        /// Returns null if this does not look like a deconstruction-declaration after all (equal sign missing).
+        ///
+        /// The syntax is either var form: `var (deconstruction-declaration, ...) = expression` or list form `(deconstruction-declaration, ...) = expression`.
+        /// </summary>
+        /// <param name="withEquals">Specifies whether to look for and consume the equals sign and the following expression.</param>
+        /// <param name="topLevel">Specifies whether to parse the terminal form of a deconstruction-declaration (which can't appear at the top-level).</param>
+        private VariableDeclarationSyntax ParseDeconstructionDeclaration(bool withEquals, bool topLevel = true)
+        {
+            Debug.Assert(topLevel || !withEquals); // withEquals can only be set at the top-level
+
+            if (this.CurrentToken.IsVar && this.PeekToken(1).Kind == SyntaxKind.OpenParenToken)
+            {
+                // parses `var (...) = expression` form
+                return ParseDeconstructionVarForm(withEquals);
+            }
+            else if (this.CurrentToken.Kind == SyntaxKind.OpenParenToken)
+            {
+                // parses `(...) = expression` form
+                return ParseDeconstructionList(withEquals, justIdentifiers: false);
+            }
+            else if (!topLevel)
+            {
+                // parses `type id` part
+                return ParseDeconstructionTypeIdPart();
+            }
+            else
+            {
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Parses `(..., ..., ...)` where each part is deconstruction variables (such as `type id` or `var (..., ...)` or another list `(..., ..., ...)`).
+        /// </summary>
+        private VariableDeclarationSyntax ParseDeconstructionList(bool withEquals, bool justIdentifiers)
+        {
+            if (this.CurrentToken.Kind == SyntaxKind.OpenParenToken)
+            {
+                var list = _pool.AllocateSeparated<VariableDeclarationSyntax>();
+
+                try
+                {
+                    var openParen = this.EatToken(SyntaxKind.OpenParenToken);
+
+                    if (this.CurrentToken.Kind != SyntaxKind.CloseParenToken)
+                    {
+                        VariableDeclarationSyntax variable;
+                        while (true)
+                        {
+                            if (justIdentifiers)
+                            {
+                                variable = ParseDeconstructionIdentifierOrIdentifiersParts(withEquals: false);
+                            }
+                            else
+                            {
+                                variable = ParseDeconstructionDeclaration(withEquals: false, topLevel: false);
+                            }
+
+                            if (variable != null)
+                            {
+                                list.Add(variable);
```

Comment:
I agree. After reviewing the new parse methods in details, it turns out most cannot return null anyways. I will try to eliminate null returns as much as possible.